### ISSUE TITLE
Bundling F# projects with C# library: Could not load file or assembly "FSharp.Core"

### ISSUE THREAD

**Issue by t-hart** [dotnet/fsharp#7426]
I have a C# class library (NuGet package) which uses an F# project within the same solution for some functionality. When trying to use the NuGet package in a different solution, I get this error:

```sh
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'FSharp.Core, Version=4.5.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
```

I am, however, unable to reproduce this when using a local version of the same package (i.e. not uploaded to an external server).

Overall, this seems like a documentation issue to me. I've spent all day reading about 'binding redirects', app.config files, implicit / explicit dependencies on `FSharp.Core` with no luck. However, I assume I have just not been able to find the right bits of documentation, and that there is a way to do this.

In theory, I could add `FSharp.Core` as a dependency to consuming applications, but that seems unnecessary and error-prone.

Also, because I'm not able to reproduce this without uploading packages to the NuGet server, testing and debugging this is taking longer than what is desirable.

In summary: C# class library / NuGet package, depends on F# project in same solution, missing `FSharp.Core` when used as an external package.

**Repro steps**

1. Create C# class library within solution
2. Depend on separate F# library in same solution
3. `dotnet pack` and publish said library to NuGet server
4. Add package dependency and get runtime exception.

**Expected behavior**

I should be able to simply use an F# project within the same solution and bundle it with the class library package. This should just work, no configuration required.

If configuration is required, I'd expect to find documentation outlining a clear set of steps to take to make this work.

**Actual behavior**

It seems necessary to configure dependency versions, binding redirects, etc., but I have not been able to make it work.

A lot of documentation has been found, but it's all scattered around, and I have been unable to find anything targeting this specific case.

**Related information**

Provide any related information (optional):

* Operating system: macOS
* .NET Runtime kind (.NET Core, .NET Framework, Mono): .NET Core
* Editing Tools (e.g. Visual Studio Version, Visual Studio: CLI (Fish shell), Emacs

**Note**
Apologies if this is not the place for it, but I don't know where else to turn and really can't seem to get anywhere with this. An option would be to rewrite the library in C#, but then I'd lose all the neat features of F# that made me want to use it in the first place.

**Comment by auduchinok** [comment]
According to these notes you should bundle `FSharp.Core` and add binding redirects for it if multiple FSharp.Core versions might be referenced by different parts of the application.
https://fsharp.github.io/2015/04/18/fsharp-core-notes.html#always-deploy-fsharpcore-as-part-of-a-compiled-application

**Comment by t-hart** [comment]
Thanks! Yeah, I saw that, but I can't figure out how to make it work. Can you help me?

This is my .fsproj file for the F# project. I've tried with and without the `DisableImplicitFSharpCoreReference`.
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netstandard20</TargetFramework>
    <DisableImplicitFSharpCoreReference>true</DisableImplicitFSharpCoreReference>
  </PropertyGroup>

  <ItemGroup>
    //...
  </ItemGroup>

  <ItemGroup>
    <PackageReference Update="FSharp.Core" Version="4.5.2" />
  </ItemGroup>

</Project>
```

And I've tried using the same PackageReference in the NuGet project's .csproj file too (and some combinations of the two), but to no avail.

I've also got an app.config file:
```xml
 <?xml version="1.0" encoding="utf-8" ?>
    <configuration>
        <runtime>
          <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
            <dependentAssembly>
              <assemblyIdentity name="FSharp.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
              <bindingRedirect oldVersion="0.0.0.0-4.5.2.0" newVersion="4.5.2.0" />
            </dependentAssembly>
          </assemblyBinding>
        </runtime>
    </configuration>

```

**Comment by auduchinok** [comment]
The `FSharp.Core` assembly from 4.5.2 package has `4.5.0.0` assembly version, thus it's the version to redirect to if you're bundling it in the application.

**Comment by t-hart** [comment]
Okay, cool. How do I bundle it, though? I've never done much bundling in .NET before, so I'm a bit inexperienced, I'm afraid.

**Comment by t-hart** [comment]
Also, it does say:
"Do _not_ bundle FSharp.Core with a library", which I find rather confusing, because right after that it tells you to bundle it with your application, and it also says that a library is just an application with a .dll extension. It's not clear whether a package like this would be a library and thus shouldn't bundle FSharp.Core or if it's an 'addin' and classified as an application, though I would lean towards the former.

**Comment by dsyme** [comment]
> a library is just an application with a .dll extension

I just checked and couldn't spot where it says this?

>  It's not clear whether a package like this would be a library and thus shouldn't bundle FSharp.Core or if it's an 'addin' and classified as an application, though I would lean towards the former.

If it's a nuget package where the effect of referencing the package is to add a reference to one or more DLLs from the package then it's a library. In this case you shouldn't bundle FSharp.Core - the nuget package should instead list a dependency on FSharp.Core.  Listing this dependency should I think be handled by .NET Core tooling (if that's not the case then something is wrong.)

If it's a nuget package that is a tool or some other executable thing (not generally in the "lib" folder) then it can bundle FSharp.Core and all the other things necessary to make a bundled application.

**Comment by cartermp** [comment]
@t-hart is it possible to share your project, or a re-creation of it that elides any sensitive info?

**Comment by t-hart** [comment]
> > a library is just an application with a .dll extension
> 
> I just checked and couldn't spot where it says this?

Yeah, you're right. Sorry, I must've misread that terribly yesterday. It says "Libraries are just ordinary .dll components (excluding those above which are applications)."


> If it's a nuget package where the effect of referencing the package is to add a reference to one or more DLLs from the package then it's a library. In this case you shouldn't bundle FSharp.Core - the nuget package should instead list a dependency on FSharp.Core. Listing this dependency should I think be handled by .NET Core tooling (if that's not the case then something is wrong.)

The listing of dependencies: is that done through `PackageReferences`? If so, do I list that dependency in the F# project or in the C# project (which is the actual package)?

In my case, it's very much a library.

> is it possible to share your project, or a re-creation of it that elides any sensitive info?

I wanted to make a repro case right now, but when I created a new package, I wasn't able to get the same result when importing it into a new project. What's more, even when importing the package I'm having problems with in another project, that seemed to work just fine in the fresh one. I'm probably doing something wrong, but I'm not sure what.

I'll do some more investigation and get back to you. Thanks for all the help and apologies for being a bit brash earlier.

**Comment by t-hart** [comment]
Alright, update time! I feel like I might've missing a key piece of the puzzle for a while.

After troubleshooting it a bit this morning, I noticed that it couldn't find the exact version of the package I was looking for, and instead used an older version that I had created locally with a higher version number. It might have been doing that for a while and I think that might have been the source of my troubles.

After clearing my caches (`nuget locals all -clear`) and reinstalling the right version of the package, things seem to be working and I'm not able to break it anymore.

So if that was indeed it, it might have been solved a while back, I just wasn't able to verify it, which also means that I don't know what fixed it. 

Anyway, I'll just verify that everything works as it should and I'll come back and close it unless something else pops up. Again, thanks!
